File: backend/app/services/hospital_os/capacity_model.py
from __future__ import annotations

from typing import Any


def _field_value(field: Any, fallback: Any = 0) -> Any:
    if isinstance(field, dict) and "value" in field:
        value = field.get("value", fallback)
        if value is None:
            return fallback
        return value
    return field if field is not None else fallback


def _numeric_field(field: Any, fallback: int = 0) -> int:
    value = _field_value(field, fallback)
    try:
        return int(round(float(value)))
    except (TypeError, ValueError):
        return fallback


def compute_surge_capacity(profile: dict[str, Any]) -> dict[str, Any]:
    capacity = profile.get("capacity", {})
    nurses = profile.get("nurses", {})

    total_beds = _numeric_field(capacity.get("total_beds"), 180)
    icu_beds = _numeric_field(capacity.get("icu_beds"), max(12, round(total_beds * 0.1)))
    occupancy = _numeric_field(capacity.get("occupancy"), 72)
    available_beds = max(0, round(total_beds * (1 - occupancy / 100)))
    er_nurses = nurses.get("er_nurses", {}).get("count", 0)
    icu_nurses = nurses.get("icu_nurses", {}).get("count", 0)

    critical_slots = max(2, round(icu_beds * (1 - occupancy / 100) + icu_nurses * 0.4 + er_nurses * 0.25))
    moderate_slots = max(4, round(available_beds * 0.32))
    minor_slots = max(8, round(available_beds * 0.42))

    return {
        "critical_slots": critical_slots,
        "moderate_slots": moderate_slots,
        "minor_slots": minor_slots,
        "total_beds": total_beds,
        "icu_beds": icu_beds,
        "available_beds": available_beds,
        "occupancy_pct": occupancy,
        "er_nurses": er_nurses,
        "icu_nurses": icu_nurses,
        "methodology": "Derived from licensed bed count, ICU capacity, occupancy, and nurse staffing.",
    }


def compute_pressure_score(incoming: dict[str, int], surge: dict[str, Any]) -> int:
    ratios = [
        incoming.get("critical", 0) / max(1, surge["critical_slots"]),
        incoming.get("moderate", 0) / max(1, surge["moderate_slots"]),
        incoming.get("minor", 0) / max(1, surge["minor_slots"]),
    ]
    return min(100, round((sum(ratios) / 3) * 100))


def run_capacity_simulation(profile: dict[str, Any], incoming: dict[str, int]) -> dict[str, Any]:
    surge = compute_surge_capacity(profile)
    capacity = {
        "critical": surge["critical_slots"],
        "moderate": surge["moderate_slots"],
        "minor": surge["minor_slots"],
        "total": surge["critical_slots"] + surge["moderate_slots"] + surge["minor_slots"],
    }

    critical_ok = incoming.get("critical", 0) <= capacity["critical"]
    moderate_ok = incoming.get("moderate", 0) <= capacity["moderate"]
    minor_ok = incoming.get("minor", 0) <= capacity["minor"]
    can_handle = critical_ok and moderate_ok and minor_ok

    overflow = {
        "critical": max(0, incoming.get("critical", 0) - capacity["critical"]),
        "moderate": max(0, incoming.get("moderate", 0) - capacity["moderate"]),
        "minor": max(0, incoming.get("minor", 0) - capacity["minor"]),
    }

    pressure_score = compute_pressure_score(incoming, surge)
    verdict = (
        "Incoming load fits within modeled hospital surge capacity."
        if can_handle
        else "Incoming load exceeds modeled surge capacity. Consider rerouting overflow patients."
    )

    if not can_handle:
        parts = []
        if overflow["critical"]:
            parts.append(f"{overflow['critical']} critical")
        if overflow["moderate"]:
            parts.append(f"{overflow['moderate']} moderate")
        if overflow["minor"]:
            parts.append(f"{overflow['minor']} minor")
        routing_advice = (
            f"Incoming load exceeds surge capacity ({surge['available_beds']} beds available at "
            f"{surge['occupancy_pct']}% occupancy). Reroute {', '.join(parts)} patients to partner facilities."
        )
    elif pressure_score >= 75:
        routing_advice = "Hospital can absorb this load but at high pressure — activate surge staffing and pre-alert ICU."
    else:
        routing_advice = (
            f"Hospital has capacity for this load with "
            f"{max(0, capacity['critical'] - incoming.get('critical', 0))} critical, "
            f"{max(0, capacity['moderate'] - incoming.get('moderate', 0))} moderate, and "
            f"{max(0, capacity['minor'] - incoming.get('minor', 0))} minor slots remaining."
        )

    return {
        "can_handle": can_handle,
        "verdict": verdict,
        "critical_ok": critical_ok,
        "moderate_ok": moderate_ok,
        "minor_ok": minor_ok,
        "pressure_score": pressure_score,
        "total_incoming": incoming.get("critical", 0) + incoming.get("moderate", 0) + incoming.get("minor", 0),
        "capacity": capacity,
        "overflow": overflow,
        "routing_advice": routing_advice,
        "surge": surge,
    }

File: frontend/src/utils/capacityModel.ts
import type { HospitalIntelligenceProfile } from '../types/hospitalIntelligence'
import { numericField } from './hospitalOsUtils'

export interface SurgeCapacity {
  critical_slots: number
  moderate_slots: number
  minor_slots: number
  total_beds: number
  icu_beds: number
  available_beds: number
  occupancy_pct: number
  er_nurses: number
  methodology: string
}

export interface CommandCenterStatus {
  capacity: SurgeCapacity
  scenario_load: { critical: number; moderate: number; minor: number; total: number }
  remaining_slots: { critical: number; moderate: number; minor: number }
  status: 'open' | 'strained' | 'saturated'
  routing_advice: string
  evidence_notes: string[]
}

export function computeSurgeCapacity(profile: HospitalIntelligenceProfile): SurgeCapacity {
  const totalBeds = numericField(profile.capacity.total_beds, 180)
  const icuBeds = numericField(profile.capacity.icu_beds, Math.max(12, Math.round(totalBeds * 0.1)))
  const occupancy = numericField(profile.capacity.occupancy, 72)
  const availableBeds = Math.max(0, Math.round(totalBeds * (1 - occupancy / 100)))
  const erNurses = profile.nurses.er_nurses.count
  const icuNurses = profile.nurses.icu_nurses.count

  const criticalSlots = Math.max(
    2,
    Math.round(icuBeds * (1 - occupancy / 100) + icuNurses * 0.4 + erNurses * 0.25)
  )
  const moderateSlots = Math.max(4, Math.round(availableBeds * 0.32))
  const minorSlots = Math.max(8, Math.round(availableBeds * 0.42))

  return {
    critical_slots: criticalSlots,
    moderate_slots: moderateSlots,
    minor_slots: minorSlots,
    total_beds: totalBeds,
    icu_beds: icuBeds,
    available_beds: availableBeds,
    occupancy_pct: occupancy,
    er_nurses: erNurses,
    methodology: 'Derived from licensed bed count, ICU capacity, current occupancy, and nurse staffing — not scenario assignment.'
  }
}

export function buildCommandCenterStatus(profile: HospitalIntelligenceProfile): CommandCenterStatus {
  const capacity = computeSurgeCapacity(profile)
  const assignment = profile.capacity.scenario_assignment
  const scenarioLoad = {
    critical: numericField(assignment?.critical_patients, 0),
    moderate: numericField(assignment?.non_critical_patients, 0),
    minor: Math.max(0, numericField(assignment?.total_patients, 0) - numericField(assignment?.critical_patients, 0) - numericField(assignment?.non_critical_patients, 0)),
    total: numericField(assignment?.total_patients, 0)
  }

  const remaining = {
    critical: Math.max(0, capacity.critical_slots - scenarioLoad.critical),
    moderate: Math.max(0, capacity.moderate_slots - scenarioLoad.moderate),
    minor: Math.max(0, capacity.minor_slots - scenarioLoad.minor)
  }

  const loadRatio =
    (scenarioLoad.critical / Math.max(1, capacity.critical_slots) +
      scenarioLoad.moderate / Math.max(1, capacity.moderate_slots) +
      scenarioLoad.minor / Math.max(1, capacity.minor_slots)) /
    3

  const status: CommandCenterStatus['status'] =
    loadRatio >= 1 ? 'saturated' : loadRatio >= 0.75 ? 'strained' : 'open'

  const eta = numericField(assignment?.eta_minutes, 0)
  const congestion = String(assignment?.congestion?.value ?? '')

  let routing_advice: string
  if (status === 'saturated') {
    routing_advice = `Current scenario load (${scenarioLoad.total} patients) exceeds modeled surge capacity. Reroute new critical patients first; moderate overflow to partner facilities within ${eta || 15} min drive radius.`
  } else if (status === 'strained') {
    routing_advice = `Hospital can absorb assigned load but with limited headroom (${remaining.critical} critical, ${remaining.moderate} moderate slots remaining). Pre-alert ICU and imaging.${congestion ? ` Route congestion: ${congestion}.` : ''}`
  } else {
    routing_advice = `Hospital has capacity for assigned scenario load with ${remaining.critical} critical, ${remaining.moderate} moderate, and ${remaining.minor} minor slots available.`
  }

  return {
    capacity,
    scenario_load: scenarioLoad,
    remaining_slots: remaining,
    status,
    routing_advice,
    evidence_notes: [
      `${capacity.total_beds} licensed beds at ${capacity.occupancy_pct}% occupancy → ~${capacity.available_beds} beds available`,
      `ICU: ${capacity.icu_beds} beds · ER nursing: ${capacity.er_nurses} FTE`,
      'Capacity model follows AHA bed surge planning guidance (not individual patient assignment counts)'
    ]
  }
}

export function computePressureScore(
  incoming: { critical: number; moderate: number; minor: number },
  capacity: SurgeCapacity
): number {
  const ratios = [
    incoming.critical / Math.max(1, capacity.critical_slots),
    incoming.moderate / Math.max(1, capacity.moderate_slots),
    incoming.minor / Math.max(1, capacity.minor_slots)
  ]
  return Math.min(100, Math.round((ratios.reduce((a, b) => a + b, 0) / 3) * 100))
}
